beacon_node/network/src/beacon_processor/worker/gossip_methods.rs

```diff
@@ -109,82 +223,194 @@ impl<T: BeaconChainTypes> Worker<T> {
+    ) {
+        let attestations_and_subnets = packages
+            .iter()
+            .map(|package| (package.attestation.as_ref(), Some(package.subnet_id)));
+
+        let results = match self
             .chain
-            .verify_unaggregated_attestation_for_gossip(attestation, Some(subnet_id))
+            .batch_verify_unaggregated_attestations_for_gossip(attestations_and_subnets)
         {
-            Ok(attestation) => attestation,
-            Err((e, attestation)) => {
-                self.handle_attestation_verification_failure(
-                    peer_id,
-                    message_id,
-                    FailedAtt::Unaggregate {
-                        attestation: Box::new(attestation),
-                        subnet_id,
-                        should_import,
-                        seen_timestamp,
-                    },
-                    reprocess_tx,
-                    e,
+            Ok(results) => results,
+            Err(e) => {
+                error!(
+                    self.log,
+                    "Batch unagg. attn verification failed";
+                    "error" => ?e
                 );
                 return;
             }
         };
 
-        // Register the attestation with any monitored validators.
-        self.chain
-            .validator_monitor
-            .read()
-            .register_gossip_unaggregated_attestation(
-                seen_timestamp,
-                attestation.indexed_attestation(),
-                &self.chain.slot_clock,
+        // Sanity check.
+        if results.len() != packages.len() {
+            // The log is `crit` since in this scenario we might be penalizing/rewarding the wrong
+            // peer.
+            crit!(
+                self.log,
+                "Batch attestation result mismatch";
+                "results" => results.len(),
+                "packages" => packages.len(),
+            )
+        }
+
+        // Map the results into a new `Vec` so that `results` no longer holds a reference to
+        // `packages`.
+        #[allow(clippy::needless_collect)] // The clippy suggestion fails the borrow checker.
+        let results = results
+            .into_iter()
+            .map(|result| result.map(|verified| verified.into_indexed_attestation()))
+            .collect::<Vec<_>>();
+
+        for (result, package) in results.into_iter().zip(packages.into_iter()) {
+            let result = match result {
+                Ok(indexed_attestation) => Ok(VerifiedUnaggregate {
+                    indexed_attestation,
+                    attestation: *package.attestation,
+                }),
+                Err(error) => Err(RejectedUnaggregate {
+                    attestation: *package.attestation,
+                    error,
+                }),
+            };
+
+            self.process_gossip_attestation_result(
+                result,
+                package.message_id.clone(),
```

Comment:
```suggestion
                package.message_id,
```

(also part of https://github.com/michaelsproul/lighthouse/commit/fe3921d50d7f7367652e9419ce7842deb1a05f84)